pkg/yang/modules.go

```diff
@@ -28,16 +28,27 @@ type Modules struct {
 	includes   map[*Module]bool   // Modules we have already done include on
 	byPrefix   map[string]*Module // Cache of prefix lookup
 	byNS       map[string]*Module // Cache of namespace lookup
+	typeDict   *typeDictionary    // Cache for type definitions.
 }
 
 // NewModules returns a newly created and initialized Modules.
 func NewModules() *Modules {
+	d := &typeDictionary{dict: map[Node]map[string]*Typedef{}}
+	// taking the lock should not be needed as this type should only be modified
+	// during init.
+	typeDict.mu.Lock()
+	defer typeDict.mu.Unlock()
+	for k, v := range typeDict.dict {
+		d.dict[k] = v
+	}
```

Comment:
I wonder if it might be less confusing to have a package default `Modules` variable, rather than a package default `typeDict` as we have now, and a separate copy in each instance of `Modules`?

This way, every time a reference is made to a `typeDict`, it is being made to a specific `(*Modules).typeDict`.

e.g.,
```go
func init() {
  defaultModules = NewModules()
}
```

and `NewModules` would include the initTypes call to the top level statement, e.g., something like

```go
initTypes(reflect.TypeOf(&meta{}), xyz)
```

Though, this would mean refactoring much of your change to refer to a `Modules` rather than a `typeDict`...

Given this might change the overall structure of your change, I'll leave the review here for the moment to get your input and the other maintainers' input on the factoring.

Reply:
I think putting the global cache objects into a global/default `Modules` struct might not be necessary because I'm not sure it serves any purpose.

Once we manage to put all the global cache objects into the `Modules` struct, each `Modules` is going to be isolated from each other, as well as from the global/default cache objects. Now, these global/default cache objects will only be populated and accessed by lower-level calls (i.e. `Parse`, `BuildAST`, `ToEntry`), so even if we have a global/default `Modules` object, the user will never have access to it via any of these calls, and therefore, its fields, like `Modules`, `SubModules`, etc. will not be populated either. So I think just keeping the global/default cache objects might be enough in the longterm if we decide to keep the lower-level calls for backwards compatibility.